pkg/util/push/request.go

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package push
+
+import (
+	"context"
+	"net/http"
+	"sync"
+
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+type Request struct {
+	cleanups []func()
+
+	httpReq                      *http.Request
+	maxMessageSize               int
+	allowSkipLabelNameValidation bool
+	parser                       ParserFunc
+	parseRequestOnce             sync.Once
+	parsedRequest                *mimirpb.WriteRequest
+	parseError                   error
+}
+
+func (r *Request) parseRequest(ctx context.Context) {
+	bufHolder := bufferPool.Get().(*bufHolder)
+	var req mimirpb.PreallocWriteRequest
+	buf, err := r.parser(ctx, r.httpReq, r.maxMessageSize, bufHolder.buf, &req)
+	if err != nil {
+		bufferPool.Put(bufHolder)
+		r.parseError = httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+		return
+	}
+	// If decoding allocated a bigger buffer, put that one back in the pool.
+	if len(buf) > len(bufHolder.buf) {
```

Comment:
`bufHolder.buf` is a slice and not a pointer to a slice so its length property can't get updated by `r.parser` and at this point must still be set to whatever it was when it came out of the pool. For what you're trying to check here, wouldn't it be better to compare the `cap` instead of the `len`?

Reply:
I just saw that `snappy.Decode` (used by `util.ParseProtoReader`) actually seems to look at the length of the given destination slice instead of the capacity, which I think is strange. Maybe we it would be better if we'd set `bufHolder.buf = bufHolder.buf[:cap(bufHolder.buf)]` before passing it into `r.parser`? That still feels strange to me though.

Reply:
I noticed that, but snappy never returns a slice that has smaller length than capacity _and_ is not the same slice that was passed to it. If snappy returns a different slice it is because the original slice was too small and it allocated a bigger one. That's how it's been working so far AFAICT.

The capacity reset makes sense to me as a precaution. It's slightly out of scope of this PR, but I will add it

Reply:
So I pushed 0192f40ab1dc5ddf018f8e1ae90968570a1fffcd which resets the capacity once it comes back from the parser. This way when you get a buffer from the pool you know that it `len(buf) == cap(buf)`. It has either been created via the bufferPool's factory which uses `make([]byte, N)` or has been maxed out by `buf = buf[:cap(buf)]`

Reply:
Thanks!